core-tools.markdown

```diff
@@ -0,0 +1,345 @@
+
+```sh
+➜ ghc -Wall hello.hs
+[1 of 1] Compiling Main             ( hello.hs, hello.o )
+
+hello.hs:1:1: warning: [-Wmissing-signatures]
+    Top-level binding with no type signature: main :: IO ()
+  |
+1 | main = putStrLn "Hello, Haskell!"
+  | ^^^^
+Linking hello ...
+```
+
+GHC has successfully compiled our program, but it has also emitted a warning about
+not annotating `main` with a type signature. While Haskell can infer
+the types of most expressions, it is recommended that top-level definitions
+are annotated with their types.
+
+We can remedy that by adding the following line above our main definition:
+
+```hs
+main :: IO ()
+```
+
+Now our `hello.hs` source file looks like this:
+
+```hs
+main :: IO ()
+main = putStrLn "Hello, world!"
+```
+
+And now (after deleting the build artifact files) GHC will compile `hello.hs`
+without warnings.
+
+`-Wall` emits many useful warnings that can easily be bugs,
+including warnings about name shadowing, unused variables, and more.
+
+Note that `-Wall` does not stop GHC from compiling your program.
+If this is something you'd like to do, specifying the flag `-Werror`
+will convert warnings to errors and will halt compilation when a warning is emitted.
+
+#### Optimisations
+
+Another very useful flag is `-O`. This flag asks GHC to compile the program with optimisations and code improvements at the cost of taking longer to compile.
+This will often make the program run much faster and perform more aggressive optimisations such as inlining and specialisation.
+For even more optimisations that may take significantly longer, `-O2` is also available.
+
+## An interactive environment
+
+GHC provides an interactive environment in a form of a
+Read-Evaluate-Print Loop (REPL) called GHCi.
+To enter the environment run the program `ghci`.
+
+```sh
+➜ ghci
+GHCi, version 9.0.1: https://www.haskell.org/ghc/  :? for help
+ghci> 
+```
+
+It provides an interactive prompt where Haskell expressions can be written and
+evaluated.
+
+For example:
+
+```sh
+ghci> 1 + 1
+2
+ghci> putStrLn "Hello, world!"
+Hello, world!
+```
+
+We can define new names:
+
+```sh
+ghci> double x = x + x
+ghci> double 2
+4
+```
+
+We can import Haskell source files using the `:load` command:
+
+```sh
+ghci> :load hello.hs
+[1 of 1] Compiling Main             ( hello.hs, interpreted )
+Ok, one module loaded.
+ghci> main
+Hello, Haskell!
+```
+
+As well as import library modules:
+
+```sh
+ghci> import Data.Bits
+ghci> shiftL 32 1
+64
+ghci> clearBit 33 0
+32
+```
+
+We can even ask what the type of an expression is using the `:type` command:
+
+```sh
+λ> :type putStrLn
+putStrLn :: String -> IO ()
+```
+
+To exit `ghci`, use the `:quit` command (or `:q` for short)
+
+```sh
+ghci> :quit
+Leaving GHCi.
+```
+
+A more thorough introduction to GHCi can be found in the [GHC user guide](https://downloads.haskell.org/~ghc/9.0.1/docs/html/users_guide/ghci.html).
+
+### Using external packages in ghci
```

Comment:
Again I wonder if we should maybe just presume one of cabal or stack and stick with it for the tutorial rather than trying to show it both ways.

Reply:
Agreed, this is the opportunity to unify an approach to both Cabal & Stack.

Reply:
My 5c: it is really annoying to have to jump over every other sentence in a section because it mentions "the other" tool that I don't use. Please, don't do that. I really hope that this page, if it covers both tools, does this in **separate (sub)sections** even if these sections are small.

Reply:
I don't think you'll experience such annoyances in this article because this happens exactly once in the article in an extremely tiny section. I'm open to seeing an alternative presentation of the information though.

Reply:
It doesn't happen exactly once unfortunately: lines 13, 231, 258 all introduce tool-related topics. After reading the whole document I don't have better suggestions though, so I'd leave it as it is probably but add the cabal-related solution to the issue raised on line 258 (see my comment there).